### PR TITLE
samples: net: echo-server: Add overlay config support

### PR DESCRIPTION
**Pull request by jukkar**
Refactor the prj*conf files so that there is only one master
prj.conf and several overlay conf files that add / change
only minor subset of functionality.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

**Comment by jukkar** [comment]
Note that this is not very well tested as I do not have many of the hardware whose config we are tweaking here.

**Comment by rlubos** [comment]
We're sligtly deviating here from what I was recommended to do in https://github.com/zephyrproject-rtos/zephyr/pull/8342#discussion_r195428570, i.e. create `boards` subdirectory in the example and put all board-specific configuration overlays there. I guess we should agree on and stick to a single approach. IMHO having configs separated in a subdirectory makes the project a little bit cleaner.

**Comment by jukkar** [comment]
> create boards subdirectory in the example and put all board-specific configuration overlays there

I was thinking about this but in this sample, the overlays are not strictly boards. Most of them are just board + some specific config or add on board. To select this config using boards subdirectory looked cumbersome. Also for example samples/net/wpanusb uses now the overlay mechanism.

**Comment by rlubos** [comment]
Well, I'm not trying to glorify one approach over another, as both are fine to me, I just think that we should stick to one approach. So if these `overlay-*` configs are prefered, I'd have to switch to it as well in samples/net/sockets/echo_(client/server). For example, now I also have two configs for a single platform - `qemu_x86.conf` and `qemu_x86_tls.conf`.

**Comment by jukkar** [comment]
Hmm, I think there is space for both ways. So to me it makes sense to have boards/ directory for strictly board specific stuff. If there is some special config like TLS support, then that could be placed in overlay file.
So for example in this echo-server sample, I think I need to add board/ directory and place possible extra options to overlay file.

For you sockets/echo_[server|client] example, the extra TLS options could be placed in overlay-tls.conf file, and you could leave the qemu_x86.conf file to boards/ directory.

**Comment by jukkar** [comment]
Refactored the config files more. Now generic board specific code is in boards/ directory. Feature configs like OpenThread support is in separate overlay config files.

File: samples/net/echo_server/overlay-ot.conf [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+CONFIG_NET_IPV6_MLD=n
+CONFIG_NET_APP_NEED_IPV6=y
+CONFIG_NET_APP_NEED_IPV4=n
+
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
+CONFIG_NET_TX_STACK_SIZE=2048
+CONFIG_NET_RX_STACK_SIZE=3072
+
+CONFIG_NET_L2_OPENTHREAD=y
```
[the remark this answers is not in the document]
Comment (jukkar) [reply samples/net/echo_server/overlay-ot.conf]: Ok, I will investigate this more.

BTW, you can just say
`cmake -DOVERLAY_CONFIG=overlay-ot.conf -DBOARD=nrf52840_pca10056 ..`
as the board specific file and global prj.conf file are automatically selected.

**Comment by jukkar** [comment]
New version which seems to compile ok with OpenThread enabled.

File: samples/net/echo_server/overlay-ot.conf [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+CONFIG_OPENTHREAD_L2_LOG_LEVEL_INFO=y
+CONFIG_OPENTHREAD_LOG_LEVEL_INFO=y
+
+CONFIG_OPENTHREAD_CHANNEL=26
+
+# We need a bit more buffers for OT
+CONFIG_NET_PKT_RX_COUNT=50
+CONFIG_NET_PKT_TX_COUNT=50
+CONFIG_NET_BUF_RX_COUNT=50
+CONFIG_NET_BUF_TX_COUNT=50
```
Comment (rlubos) [inline samples/net/echo_server/overlay-ot.conf]: We should specifiy IPv6 addresses for Thread explicitly:
```
CONFIG_NET_APP_MY_IPV6_ADDR="fdde:ad00:beef::2"
CONFIG_NET_APP_PEER_IPV6_ADDR="fdde:ad00:beef::1"
```
I saw that kw41z didn't do it in it's config for OT, yet it's not correct even if it worked with default addresses used by the sample. Thread devices should not add addresses with prefixes that are not registered in the network, it's a spec violation.

**Review by dbkinder — APPROVED** [review]
docs LTGM

**Comment by jukkar** [comment]
Thanks @rlubos, I merged your suggested OT changes.

**Review by pfalcon — APPROVED** [review]
echo_server/echo_client configs have become augean stables long ago, so any work on cleaning up is good. Even if some issue detected ("can't test all configs"), it's better to fix them, then be afraid to touch to refactor them, like it was for some time.

**Review by rlubos — APPROVED** [review]
@jukkar Thanks for taking my remarks into consideration. Just given it one last test and OpenThread works fine. There is is a bug in `echo_server` though that was introduced yestarday, will submit a fix.

I guess `echo_client` will be next in the line? There is a small incompability now between the samples (OT channel namely).

**Comment by jukkar** [comment]
> I guess echo_client will be next in the line? There is a small incompability now between the samples (OT channel namely).

Yes, I plan to handle echo-client next.

**Pull request merged**